framework/include/materials/FunctorMaterialProperty.h

```diff
@@ -0,0 +1,225 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseMesh.h"
+#include "MooseTypes.h"
+#include "MooseError.h"
+#include "FunctorInterface.h"
+#include "Moose.h"
+#include "Limiter.h"
+#include "FVUtils.h"
+
+#include "libmesh/elem.h"
+#include "libmesh/remote_elem.h"
+#include "libmesh/tensor_tools.h"
+
+#include <unordered_map>
+#include <functional>
+
+/**
+ * A material property that is evaluated on-the-fly via calls to various overloads of \p operator()
+ */
+template <typename T>
+class FunctorMaterialProperty : public FunctorInterface<T>
+
+{
+public:
+  FunctorMaterialProperty(const std::string & name) : _name(name) {}
+
+  /**
+   * Set the functor that will be used in calls to \p evaluate overloads
+   * @param mesh The mesh that the functor is defined on
+   * @param block_ids The block/subdomain IDs that the user-provided functor is valid for
+   * @param my_lammy The functor that defines this how this object is evaluated
+   */
+  template <typename PolymorphicLambda>
+  void setFunctor(const MooseMesh & mesh,
+                  const std::set<SubdomainID> & block_ids,
+                  PolymorphicLambda my_lammy);
+
```

Comment:
I still don't see the "natural" way for a user to provide arg-specific lambda/functor implementation.  I'd really like to have a blessed way to approach that issue without having to call `typeid` or other bad solutions.  Even something as simple as returning an error for some subset of "unsupported" locations I envision being not uncommon.

Reply:
For now we ll just keep a single shared overload. It s not obvious that people will want anything else